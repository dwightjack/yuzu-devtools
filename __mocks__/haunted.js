const haunted = require('haunted/index');

module.exports = {
  ...haunted,
  useContext: jest.fn(),
};
