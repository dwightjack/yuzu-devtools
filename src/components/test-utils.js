import { render, html } from 'lit-html';

export const mockComponent = (tag) => (props) => {
  const el = document.createElement(tag);
  Object.keys(props).forEach((key) => {
    el.setAttribute(key, props[key]);
  });
  return el;
};

export function cycle() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0);
  });
}

const cleanup = (wrapper) =>
  wrapper.querySelectorAll('style').forEach((style) => {
    style.remove();
  });

const toObject = (wrapper) => ({
  $root: wrapper,
  get html() {
    return wrapper.innerHTML;
  },
  find(selector) {
    return wrapper.querySelector(selector);
  },
  findAll(selector) {
    return wrapper.querySelectorAll(selector);
  },
});

export function toHTML(template, props = {}, raw = false) {
  const wrapper = document.createElement('div');

  Object.defineProperty(wrapper, 'isRoot', {
    value: true,
    enumerable: false,
  });

  if (props && typeof template === 'function') {
    // eslint-disable-next-line no-param-reassign
    template = template(props);
  }
  if (Array.isArray(template)) {
    render(
      html`
        ${template}
      `,
      wrapper,
    );
  } else {
    render(template, wrapper);
  }
  cleanup(wrapper);
  return raw ? wrapper : toObject(wrapper);
}

export function toHTMLAsync(template, props, raw = false) {
  const root = toHTML(template, props, true);
  return cycle().then(() => {
    cleanup(root);
    return raw ? root : toObject(root);
  });
}
