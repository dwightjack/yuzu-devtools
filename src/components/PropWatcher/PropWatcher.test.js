import { fireEvent } from 'dom-testing-library';
import { virtual } from 'haunted';
import { toHTMLAsync } from '../test-utils';
import PropWatcherDefault from './PropWatcher';

const PropWatcher = virtual(PropWatcherDefault);

describe('PropWatcher', () => {
  const props = { uid: '_0', key: 'name' };
  test('matches default snapshot', async () => {
    const { html } = await toHTMLAsync(PropWatcher, props);
    expect(html).toMatchSnapshot();
  });

  test('matches checked snapshot', async () => {
    const { html } = await toHTMLAsync(PropWatcher, {
      ...props,
      watched: true,
    });
    expect(html).toMatchSnapshot();
  });

  test('trigger a "toggle" event on the root when clicking on the checkbox', async () => {
    const onToggle = jest.fn();
    const wrapper = await toHTMLAsync(PropWatcher, { ...props, watched: true });

    wrapper.$root.addEventListener('toggle', ({ detail }) =>
      onToggle(detail.watched),
    );

    fireEvent(wrapper.find('input'), new Event('click'));
    expect(onToggle).toHaveBeenCalledWith(false);
  });
});
