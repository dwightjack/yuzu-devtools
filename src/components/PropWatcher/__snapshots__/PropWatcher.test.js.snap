// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropWatcher matches checked snapshot 1`] = `
<label title="Watch changes">
  <input type="checkbox"
         checked
         value="_0:name"
  >
  <svg title="ic_visibility_24px.svg">
  </svg>
</label>
`;

exports[`PropWatcher matches default snapshot 1`] = `
<label title="Watch changes">
  <input type="checkbox"
         value="_0:name"
  >
  <svg title="ic_visibility_24px.svg">
  </svg>
</label>
`;
